test/test_jsoo_react_ml.ml

```diff
@@ -0,0 +1,778 @@
+(*       React.Dom.forwardRef (fun ~children ref_ -> *)
+(*           React.Dom.Html.(button [|ref ref_; className "FancyButton"|] children) ) *)
+(*   end in *)
+(*   withContainer (fun c -> *)
+(*       let count = ref 0 in *)
+(*       let buttonRef = *)
+(*         React.Dom.Ref.callbackDomRef (fun _ref -> count := !count + 1) *)
+(*       in *)
+(*       act (fun () -> *)
+(*           React.Dom.render *)
+(*             React.Dom.Html.(FancyButton.make ~ref:buttonRef [div [||] []]) *)
+(*             (Html.element c) ) ; *)
+(*       assert_equal !count 1 ) *)
+
+let testUseRef () =
+  let module DummyComponentWithRefAndEffect = struct
+    let%component make ~cb () =
+      let myRef = React.useRef 1 in
+      React.useEffect0 (fun () ->
+          let open React.Ref in
+          setCurrent myRef (current myRef + 1) ;
+          cb myRef ;
+          None ) ;
+      React.Dom.Html.div [||] []
+  end in
+  withContainer (fun c ->
+      let myRef = ref None in
+      let cb reactRef = myRef := Some reactRef in
+      act (fun () ->
+          React.Dom.render
+            (DummyComponentWithRefAndEffect.make ~cb ())
+            (Html.element c) ) ;
+      assert_equal
+        (myRef.contents |> Option.map (fun item -> item |> React.Ref.current))
+        (Some 2) )
+
+let testChildrenMapWithIndex () =
+  let module DummyComponentThatMapsChildren = struct
+    let%component make ~children () =
+      React.Dom.Html.(
+        div [||]
+          [ React.Children.mapWithIndex children (fun element index ->
+                React.cloneElement element
+                  (let open Js_of_ocaml.Js.Unsafe in
+                  obj [|("key", inject index); ("data-index", inject index)|]) )
+          ])
+  end in
+  withContainer (fun c ->
+      act (fun () ->
+          React.Dom.render
+            React.Dom.Html.(
+              DummyComponentThatMapsChildren.make
+                ~children:[div [||] [int 1]; div [||] [int 2]; div [||] [int 3]]
+                ())
+            (Html.element c) ) ;
+      assert_equal c##.innerHTML
+        (Js.string
+           "<div><div data-index=\"0\">1</div><div \
+            data-index=\"1\">2</div><div data-index=\"2\">3</div></div>" ) )
+
+let testFragmentModule () =
+  withContainer (fun c ->
+      act (fun () ->
+          React.Dom.render
+            React.Dom.Html.(
+              fragment [div [||] [string "Hello"]; div [||] [string "World"]])
+            (Html.element c) ) ;
+      assert_equal c##.innerHTML (Js.string "<div>Hello</div><div>World</div>") )
+
+(* NOTE: Effectively a duplicate of the previous test *)
+let testFragmentSyntax () =
+  withContainer (fun c ->
+      act (fun () ->
+          React.Dom.render
+            React.Dom.Html.(
+              fragment [div [||] [string "Hello"]; div [||] [string "World"]])
+            (Html.element c) ) ;
+      assert_equal c##.innerHTML (Js.string "<div>Hello</div><div>World</div>") )
```

Comment:
There's really no difference between fragment "module" and "syntax". There might eventually be though, so keeping it around for now.

Reply:
Right, this only made sense under the Reason syntax.

Reply:
Removed `testFragmentSyntax`